src/components/ha-icon.ts

```diff
@@ -1,34 +1,91 @@
 import "@polymer/iron-icon/iron-icon";
-import type { IronIconElement } from "@polymer/iron-icon/iron-icon";
-import { Constructor } from "../types";
+import { get, Store } from "idb-keyval";
+import {
+  customElement,
+  LitElement,
+  property,
+  PropertyValues,
+  html,
+  TemplateResult,
+} from "lit-element";
+import "./ha-svg-icon";
 
-const ironIconClass = customElements.get("iron-icon") as Constructor<
-  IronIconElement
->;
+const iconStore = new Store("hass-icon-db", "mdi-icon-store");
+const partPromise = {};
+const MDI_PREFIXES = ["mdi", "hass", "hassio"];
 
-let loaded = false;
+@customElement("ha-icon")
+export class HaIcon extends LitElement {
+  @property() public icon?: string;
 
-export class HaIcon extends ironIconClass {
-  private _iconsetName?: string;
+  @property() private _path?: string;
 
-  public listen(
-    node: EventTarget | null,
-    eventName: string,
-    methodName: string
-  ): void {
-    super.listen(node, eventName, methodName);
+  @property() private _noMdi = false;
 
-    if (!loaded && this._iconsetName === "mdi") {
-      loaded = true;
-      import(/* webpackChunkName: "mdi-icons" */ "../resources/mdi-icons");
+  protected updated(changedProps: PropertyValues) {
+    if (changedProps.has("icon")) {
+      this._loadIcon();
     }
   }
-}
 
+  protected render(): TemplateResult {
+    if (!this.icon) {
+      return html``;
+    }
+    if (this._noMdi) {
+      return html`<iron-icon .icon=${this.icon}></iron-icon>`;
+    }
+    return html`<ha-svg-icon .path=${this._path}></ha-svg-icon>`;
+  }
+
+  private async _loadIcon() {
+    if (!this.icon) {
+      return;
+    }
+    const icon = this.icon.split(":", 2);
+    if (!MDI_PREFIXES.includes(icon[0])) {
+      this._noMdi = true;
+      return;
+    }
+
+    this._noMdi = false;
+
+    const iconName = icon[1];
+    const cachedPath: string = await get(iconName, iconStore);
+    if (cachedPath) {
+      this._path = cachedPath;
+      return;
+    }
+    const part = iconName[0];
+    if (part in partPromise) {
+      this._setPath(partPromise[part], iconName);
+      return;
+    }
+    partPromise[part] = fetch(`/static/mdi/${part}.json`).then((response) =>
+      response.json()
+    );
+    this._setPath(partPromise[part], iconName, true);
```

Comment:
Since awaiting promises doesn't make extra copies of objects, let's extract saving from `setPath` responsibilities.

```ts
this._setPath(…)
this._writeCache(…)
```